tracer/src/Datadog.Trace/AppSec/AppSecRequestContext.cs

```diff
@@ -47,27 +49,38 @@ internal void CloseWebSpan(TraceTagCollection tags, Span span)
                 }
             }
 
-            if (_raspStackTraces?.Count > 0)
+            if (_wafTimeout > 0)
             {
-                span.SetMetaStruct(StackKey, MetaStructHelper.ObjectToByteArray(_raspStackTraces));
+                span.Tags.SetMetric(Metrics.WafTimeouts, _wafTimeout);
             }
 
             if (_wafError != null)
             {
-                tags.SetTag(Tags.WafError, _wafError.ToString());
+                span.Tags.SetMetric(Metrics.WafError, _wafError);
             }
 
             if (_wafRaspError != null)
             {
-                tags.SetTag(Tags.RaspWafError, _wafRaspError.ToString());
+                span.Tags.SetMetric(Metrics.RaspWafError, _wafRaspError);
+            }
+
+            if (_raspStackTraces?.Count > 0)
+            {
+                span.SetMetaStruct(StackKey, MetaStructHelper.ObjectToByteArray(_raspStackTraces));
             }
 
-            _raspTelemetryHelper?.GenerateRaspSpanMetricTags(span.Tags);
+            _raspMetricsHelper?.GenerateRaspSpanMetricTags(span.Tags);
         }
     }
 
-    internal void CheckWAFError(int code, bool isRasp)
+    internal void CheckWAFError(IResult result, bool isRasp)
     {
+        if (result.Timeout)
+        {
+            Interlocked.Increment(ref _wafTimeout);
```

Comment:
I'm a little unsure about the concurrency expectations in this class - the interlocked here implies it's meant to be able to be called from multiple threads, but basically nothing else in the type is thread safe as far as I can tell 😅

Reply:
Better safe than sorry, but you are right